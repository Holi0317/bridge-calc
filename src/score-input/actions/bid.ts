import {IPlayerMap} from '../../types'

export const BID: 'CURRENT_GAME/BID' = 'CURRENT_GAME/BID'
export interface IBidAction {
  type: typeof BID,
  payload?: IPlayerMap<number>
}

/**
 * End bidding process in bridge game
 * @param payload - A map that maps player ID to their bid choice.
 * If undefined, bid property in currentGame state will be used
 */
export function bid(payload?: IPlayerMap<number>): IBidAction {
  return {type: BID, payload}
}
