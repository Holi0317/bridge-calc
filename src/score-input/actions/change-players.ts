import {IPlayerMap} from '../../types'

export const CHANGE_PLAYERS: 'CURRENT_GAME/CHANGE_PLAYERS' = 'CURRENT_GAME/CHANGE_PLAYERS'
export interface IChangePlayersAction {
  type: typeof CHANGE_PLAYERS,
  newNames: IPlayerMap<string>,
  maker: string,
  rounds: number,
  time: Date
}

/**
 * Change players of current game.
 * Change can include: Change name, add player or remove player.
 * After this action is dispatched, the stage will revert to waitingBid if original stage is waitingWin. (To avoid invalid bid after changing player)
 *
 * This action can also be used to change maker. Just pass in original name map, original rounds and specify new maker in `maker` property.
 *
 * This action can also be used to change number of rounds. Similar to change maker, only update rounds property.
 *
 * @param newNames - New PlayerMap for the names.
 * Order of this map represent the new player order.
 * If new player is added, generate a new unique ID for him/her.
 *
 * @param maker - As order may have changed, new maker must be selected.
 * This property represent the ID of new maker.
 * The player order will be sorted by this property and newNames property.
 * This can also be used to change maker.
 *
 * @param rounds - Number of rounds after dispatching this action.
 * If the given rounds is less than current round, the game will end immediately.
 * If the above case happens, endTime of the game will be starting time + 1 min.
 * Because providing such round is invalid and validation should not be done in reducer.
 */
export function changePlayers(newNames: IPlayerMap<string>, maker: string, rounds: number): IChangePlayersAction {
  return {
    type: CHANGE_PLAYERS,
    newNames,
    maker,
    rounds,
    time: new Date()
  }
}
