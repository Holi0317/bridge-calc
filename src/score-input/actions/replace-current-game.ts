import { createAction } from "typesafe-actions";
import { GameState } from "../reducer";
import { ActionTypes } from "../../action-types";

export const replaceCurrentGameAction = createAction(
  ActionTypes.REPLACE_CURRENT_GAME,
  (payload: GameState) => payload
)();
