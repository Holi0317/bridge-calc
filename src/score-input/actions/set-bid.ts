import {IPlayerMap} from '../../types'

export const SET_BID: 'CURRENT_GAME/SET_BID' = 'CURRENT_GAME/SET_BID'
export interface ISetBidAction {
  type: typeof SET_BID,
  payload: IPlayerMap<number>
}

/**
 * Set the bid property in current game cache.
 * Not to be confused with BID action.
 *
 * @param payload - A map that maps player ID to their bid choice
 */
export function setBid(payload: IPlayerMap<number>): ISetBidAction {
  return {type: SET_BID, payload}
}
