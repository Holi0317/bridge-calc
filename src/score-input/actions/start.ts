import cuid from 'cuid'
import {IPlayerMap} from '../../types'
import {ActionTypes} from '../../action-types'

export interface IStartAction {
  type: ActionTypes.START,
  rounds: number,
  playerNames: IPlayerMap<string>,
  startTime: number,
  startingRound: number,
  id: string
}

/**
 * Helper function to Change player names array to object with random generated player ID as key.
 * @param playerNames
 */
function namesToMap(playerNames: string[]): IPlayerMap<string> {
  const result: IPlayerMap<string> = {}
  playerNames.forEach(name => {
    result[cuid()] = name
  })
  return result
}

/**
 * Start a new game that overrides existing game.
 *
 * @param rounds - Total number of rounds in the whole game.
 * @param playerNames - Player names in the game. Each player will get an unique ID when game starts.
 * Ordering of the names will be used as the player order for first round.
 * @param startingRound - The round to start from
 */
export function startAction(rounds: number, playerNames: string[], startingRound: number): IStartAction {
  return {
    type: ActionTypes.START,
    rounds,
    playerNames: namesToMap(playerNames),
    startTime: new Date().getTime(),
    startingRound,
    id: cuid()
  }
}
