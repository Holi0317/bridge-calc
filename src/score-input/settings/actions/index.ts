import { InitSettingsAction } from "./init-settings";
import { SetMakerAction } from "./set-maker";
import { SetNamesAction } from "./set-names";
import { AddNameAction } from "./add-name";
import { ToggleExpandAction } from "./toggle-expand";

export type GameSettingsActions =
  | InitSettingsAction
  | SetMakerAction
  | SetNamesAction
  | AddNameAction
  | ToggleExpandAction;
