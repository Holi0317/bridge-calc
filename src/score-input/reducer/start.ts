import {GameStage} from '../game-stage'
import {skip} from './skip'
import {bidWinGenerator} from './bid-win-generator'
import {fillObj} from '../../utils'

import {IStartAction} from '../../actions/current-game'
import {IWaitingBidState} from './types'

/**
 * Helper function for reducer.
 * Start a new game base on start action parameters.
 * @param action
 */
export function start(action: IStartAction) {
  const playerIDs = Object.keys(action.playerNames)

  const firstState: IWaitingBidState = {
    stage: GameStage.waitingBid,
    rounds: action.rounds,
    startTime: action.startTime,
    names: action.playerNames,
    scores: fillObj({}, playerIDs, []),
    bid: bidWinGenerator(playerIDs),
    currentPlayerOrder: playerIDs,
    currentRound: 1
  }
  return skip(firstState, action.startingRound - 1, action.startTime)
}
