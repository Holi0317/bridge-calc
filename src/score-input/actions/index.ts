import {IBidAction} from './bid'
import {IChangePlayersAction} from './change-players'
import {ISetBidAction} from './set-bid'
import {ISetWinAction} from './set-win'
import {ISkipAction} from './skip'
import {IStartAction} from './start'
import {IUndoAction} from './undo'
import {IWinAction} from './win'

export type CurrentGameActions =
  | IBidAction
  | IChangePlayersAction
  | ISetBidAction
  | ISetWinAction
  | ISkipAction
  | IStartAction
  | IUndoAction
  | IWinAction
