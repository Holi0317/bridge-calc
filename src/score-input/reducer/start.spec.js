import {start} from './start'

import {startAction as startAction} from '../actions/start'
import * as lolex from 'lolex'

const playerNames = ['John', 'Mary', 'Henry', 'Joe']

let clock = null
beforeEach(() => {
  clock = lolex.install()
})
afterEach(() => {
  if (clock) {
    clock.uninstall()
  }
})

test('it should generate state according to action', () => {
  const action = startAction(13, playerNames, 1)
  const actual = start(action)
  expect(actual).toMatchSnapshot()
})

test('Starting in the middle of the game should work', () => {
  const action = startAction(13, playerNames, 5)
  const actual = start(action)
  expect(actual).toMatchSnapshot()
})
