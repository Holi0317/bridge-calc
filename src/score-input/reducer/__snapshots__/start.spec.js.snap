// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Starting in the middle of the game should work 1`] = `
Object {
  "bid": Object {
    "player_5": 0,
    "player_6": 0,
    "player_7": 0,
    "player_8": 0,
  },
  "currentPlayerOrder": Array [
    "player_5",
    "player_6",
    "player_7",
    "player_8",
  ],
  "currentRound": 5,
  "names": Object {
    "player_5": "John",
    "player_6": "Mary",
    "player_7": "Henry",
    "player_8": "Joe",
  },
  "rounds": 13,
  "scores": Object {
    "player_5": Array [
      0,
      0,
      0,
      0,
    ],
    "player_6": Array [
      0,
      0,
      0,
      0,
    ],
    "player_7": Array [
      0,
      0,
      0,
      0,
    ],
    "player_8": Array [
      0,
      0,
      0,
      0,
    ],
  },
  "stage": "waitingBid",
  "startTime": 1970-01-01T00:00:00.000Z,
}
`;

exports[`it should generate state according to action 1`] = `
Object {
  "bid": Object {
    "player_1": 0,
    "player_2": 0,
    "player_3": 0,
    "player_4": 0,
  },
  "currentPlayerOrder": Array [
    "player_1",
    "player_2",
    "player_3",
    "player_4",
  ],
  "currentRound": 1,
  "names": Object {
    "player_1": "John",
    "player_2": "Mary",
    "player_3": "Henry",
    "player_4": "Joe",
  },
  "rounds": 13,
  "scores": Object {
    "player_1": Array [],
    "player_2": Array [],
    "player_3": Array [],
    "player_4": Array [],
  },
  "stage": "waitingBid",
  "startTime": 1970-01-01T00:00:00.000Z,
}
`;
